core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -297,17 +314,21 @@ private MatchingNodes findMatchingNodes(ShardRouting shard, RoutingAllocation al
             }
         }
 
-        return new MatchingNodes(nodesToSize);
+        assert explain || nodeDecisions.size() == 0 : "node decision map should not have entries if we are not in explain mode";
```

Comment:
you're right, this is wrong as we could have nodes that didn't go through the decision (was aborted earlier in the loop iteration), removing this